src/JSON/Utils/Views.dfy

```diff
@@ -0,0 +1,121 @@
+include "../../BoundedInts.dfy"
+
+module {:options "-functionSyntax:4"} JSON.Utils.Views.Core {
+  import opened BoundedInts
+
+  type View = v: View_ | v.Valid? witness View([], 0, 0)
+  datatype View_ = View(s: bytes, beg: uint32, end: uint32) {
+    ghost const Valid?: bool :=
+      0 <= beg as int <= end as int <= |s| < TWO_TO_THE_32;
+
+    static const Empty: View :=
+      View([], 0, 0)
+
+    const Empty? :=
+      beg == end
+
+    function Length(): uint32 requires Valid? {
+      end - beg
+    }
+
+    function Bytes(): bytes requires Valid? {
+      s[beg..end]
+    }
+
+    static function OfBytes(bs: bytes) : (v: View)
+      requires |bs| < TWO_TO_THE_32
+      ensures v.Bytes() == bs
+    {
+      View(bs, 0 as uint32, |bs| as uint32)
+    }
+
+    static function OfString(s: string) : bytes
+      requires forall c: char | c in s :: c as int < 256
+    {
+      seq(|s|, i requires 0 <= i < |s| =>
+        assert s[i] in s; s[i] as byte)
+    }
+
+    ghost predicate SliceOf?(v': View) {
+      v'.s == s && v'.beg <= beg && end <= v'.end
+    }
+
+    ghost predicate StrictPrefixOf?(v': View) {
+      v'.s == s && v'.beg == beg && end < v'.end
+    }
+
+    ghost predicate StrictSuffixOf?(v': View) {
+      v'.s == s && v'.beg < beg && end == v'.end
+    }
+
+    predicate Byte?(c: byte)
+      requires Valid?
+    {
+      Bytes() == [c]
+    } by method {
+      return Length() == 1 && At(0) == c;
+    }
+
+    predicate Char?(c: char)
+      requires Valid?
+      requires c as int < 256
+    {
+      Byte?(c as byte)
+    }
+
+    ghost predicate ValidIndex?(idx: uint32) {
+      beg as int + idx as int < end as int
+    }
+
+    function At(idx: uint32) : byte
+      requires Valid?
+      requires ValidIndex?(idx)
+    {
+      s[beg + idx]
+    }
+
+    function Peek(): (r: opt_byte)
+      requires Valid?
+      ensures r < 0 <==> Empty?
+    {
+      if Empty? then -1
+      else At(0) as opt_byte
+    }
+
+    method Blit(bs: array<byte>, start: uint32 := 0)
```

Comment:
I agree with you; Blit is a bit opaque.  But I think it should be `CopyTo`, to make it clear that the original object isn't affected.

`bs` is the convention throughout, but I agree with you that it's not the best.  However `bytes` isn't informative either; it just repeats the type.  I think the right name would be `target` or `dest`; I chose `dest` based on https://stackoverflow.com/questions/60532503/whats-different-between-words-target-and-dest

`start` I would prefer to not change, because other functions have `end` (here since it's a public API I didn't use `beg`)